src/Raven.Server/Documents/AI/Embeddings/QueryEmbeddingsBatchingWorker.cs

```diff
@@ -0,0 +1,283 @@
+        {
+            CancellationToken.ThrowIfCancellationRequested();
+            
+            var request = new QueryEmbeddingsRequest(values, cancellationToken);
+
+            _requestQueue.Enqueue(request);
+
+            _workAvailable.Set();
+
+            return request.TaskCompletionSource.Task;
+        }
+
+        private async Task WorkerLoopAsync()
+        {
+            while (CancellationToken.IsCancellationRequested == false)
+            {
+                try
+                {
+                    WaitHandle.WaitAny([_workAvailable, CancellationToken.WaitHandle]);
+
+                    CancellationToken.ThrowIfCancellationRequested();
+
+                    if (await _queueReaderLock.WaitAsync(0, CancellationToken))
+                    {
+                        List<QueryEmbeddingsRequest> requests;
+                        var totalNumberOfValues = 0;
+
+                        var mightBeMore = false;
+
+                        try
+                        {
+                            requests = new List<QueryEmbeddingsRequest>();
+
+                            while (_requestQueue.TryDequeue(out var request)) // todo: take into account model's token limit
+                            {
+                                if (request.TaskCompletionSource.Task.IsCanceled)
+                                    continue;
+
+                                requests.Add(request);
+
+                                totalNumberOfValues += request.Values.Count;
+
+                                if (requests.Count > _configuration.QueryEmbeddingsMaxBatchSize)
+                                {
+                                    mightBeMore = true;
+                                    break;
+                                }
+                            }
+                        }
+                        finally
+                        {
+                            _queueReaderLock.Release();
+                        }
+
+                        if (mightBeMore)
+                            _workAvailable.Set();
+
+                        if (requests.Count > 0)
+                            await ProcessBatchAsync(requests, totalNumberOfValues);
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    while (_requestQueue.TryDequeue(out var request))
+                        request.TaskCompletionSource.TrySetCanceled();
+                    
+                    break;
+                }
+                catch (Exception ex)
+                {
+                    if (Logger.IsErrorEnabled)
+                        Logger.Error($"Error in query embeddings batching worker for connection string '{_connectionStringIdentifier.Value}' in database '{_databaseName}'", ex);
+                }
+            }
+        }
+
+        private async Task ProcessBatchAsync(List<QueryEmbeddingsRequest> requests, int totalNumberOfValues)
```

Comment:
I think that the way we approach it is wrong. The long duration we have to deal with is the time the model will run, no?

So we need to split things betwee _initiating_ the call and actually processing.

Right now, we have N workers that compete with one another, but it is far more reasonable spawn those tasks as they come, then make sure that we aren't exceeding the max concurrency limit.

Same for hitting the actual rate limits.